flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/slotpool/BulkSlotProviderImplTest.java

```diff
@@ -0,0 +1,245 @@
+	}
+
+	@Test
+	public void testBulkSlotAllocationFulfilledWithAvailableSlots() throws Exception {
+		final PhysicalSlotRequest request1 = createPhysicalSlotRequest();
+		final PhysicalSlotRequest request2 = createPhysicalSlotRequest();
+		final List<PhysicalSlotRequest> requests = Arrays.asList(request1, request2);
+
+		addSlotToSlotPool();
+		addSlotToSlotPool();
+
+		final CompletableFuture<Collection<PhysicalSlotRequest.Result>> slotFutures = allocateSlots(requests);
+
+		final Collection<PhysicalSlotRequest.Result> results = slotFutures.get(TIMEOUT.getSize(), TIMEOUT.getUnit());
+		final Collection<SlotRequestId> resultRequestIds = results.stream()
+			.map(PhysicalSlotRequest.Result::getSlotRequestId)
+			.collect(Collectors.toList());
+
+		assertThat(resultRequestIds, containsInAnyOrder(request1.getSlotRequestId(), request2.getSlotRequestId()));
+	}
+
+	@Test
+	public void testBulkSlotAllocationFulfilledWithNewSlots() {
+		final List<PhysicalSlotRequest> requests = Arrays.asList(
+			createPhysicalSlotRequest(),
+			createPhysicalSlotRequest());
+		final CompletableFuture<Collection<PhysicalSlotRequest.Result>> slotFutures = allocateSlots(requests);
+
+		addSlotToSlotPool();
+
+		assertThat(slotFutures.isDone(), is(false));
+
+		addSlotToSlotPool();
+
+		assertThat(slotFutures.isDone(), is(true));
+		assertThat(slotFutures.isCompletedExceptionally(), is(false));
+	}
+
+	@Test
+	public void testBulkSlotAllocationTimeoutsIfUnfulfillable() {
+		final List<PhysicalSlotRequest> requests = Arrays.asList(
+			createPhysicalSlotRequest(),
+			createPhysicalSlotRequest());
+		final CompletableFuture<Collection<PhysicalSlotRequest.Result>> slotFutures = allocateSlots(requests);
+
+		addSlotToSlotPool();
+
+		assertThat(slotPool.getAllocatedSlots().listSlotInfo(), hasSize(1));
+
+		clock.advanceTime(TIMEOUT.toMilliseconds() + 1L, TimeUnit.MILLISECONDS);
+
+		try {
+			// wait util the requests timed out
+			slotFutures.get();
+			fail("Expected that the slot futures time out.");
+		} catch (Exception e) {
+			final Optional<Throwable> cause = ExceptionUtils.findThrowableWithMessage(
+				e,
+				"Slot request bulk is not fulfillable!");
+			assertThat(cause.isPresent(), is(true));
+			assertThat(cause.get(), instanceOf(TimeoutException.class));
+		}
+	}
+
+	@Test
+	public void testFailedBulkSlotAllocationReleasesAllocatedSlot() {
+		final List<PhysicalSlotRequest> requests = Arrays.asList(
+			createPhysicalSlotRequest(),
+			createPhysicalSlotRequest());
+		final CompletableFuture<Collection<PhysicalSlotRequest.Result>> slotFutures = allocateSlots(requests);
+
+		addSlotToSlotPool();
+
+		assertThat(slotPool.getAllocatedSlots().listSlotInfo(), hasSize(1));
+
+		clock.advanceTime(TIMEOUT.toMilliseconds() + 1L, TimeUnit.MILLISECONDS);
+
+		try {
+			// wait util the requests timed out
+			slotFutures.get();
+			fail("Expected that the slot futures time out.");
+		} catch (Exception e) {
+			// expected
+		}
```

Comment:
Could this be a private helper method `Exception  allocateSlotsAndWaitForTimeout()` to deduplicate with other tests?
```
private Exception allocateSlotsAndWaitForTimeout() {
		final List<PhysicalSlotRequest> requests = Arrays.asList(
			createPhysicalSlotRequest(),
			createPhysicalSlotRequest());
		final CompletableFuture<Collection<PhysicalSlotRequest.Result>> slotFutures = allocateSlots(requests);

		addSlotToSlotPool();

		assertThat(slotPool.getAllocatedSlots().listSlotInfo(), hasSize(1));

		clock.advanceTime(TIMEOUT.toMilliseconds() + 1L, TimeUnit.MILLISECONDS);

		try {
			// wait util the requests timed out
			slotFutures.get();
		} catch (Exception e) {
			// expected
			return e;
		}
		fail("Expected that the slot futures time out.");
		return new Exception("Unexpected");
	}
```

Reply:
done.